packages/pass-style/src/types.js

```diff
@@ -6,19 +6,21 @@ export {};
  * } PrimitiveStyle
  */
 
+/** @typedef {'copyRecord' | 'copyArray' | 'tagged'} ContainerStyle */
+
 /**
  * @typedef { PrimitiveStyle |
- *   'copyRecord' | 'copyArray' | 'tagged' |
+ *   ContainerStyle |
  *   'remotable' |
  *   'error' | 'promise'
  * } PassStyle
  */
 
-// TODO declare more precise types throughout this file, so the type system
-// and IDE can be more helpful.
+/** @typedef {import('type-fest').Primitive | CopyArray | CopyRecord | CopyTagged } PassByCopy */
+/** @typedef {RemotableObject | Promise<RemotableObject> | Promise<PassByCopy> } PassByRef */
 
 /**
- * @typedef {any} Passable
+ * @typedef {PassByCopy | PassByRef} Passable
```

Comment:
I can't fit this recursive definition into JSDoc, but things like `PureData` suggest that `Passable` should be parameterized...

```suggestion
 * @template [C=PassableCap]
 * @template [E=Error]
 * @typedef {import('type-fest').Primitive | Container<Passable<C, E>> | C | E} Passable
```

where...

```js
/**
 * @template [C=PassableCap]
 * @template [E=Error]
 * @typedef {CopyArray<Passable<C, E>> | CopyRecord<Passable<C, E>> | CopyTagged<Passable<C, E>>} Container
 */
```

and then:

```js
 /**
  * @typedef {Passable<never, never>} PureData
  */
```

Reply:
I ran into circular definition errors. I'll try more. https://stackoverflow.com/a/70501728/46040 has a tip.

I also want to try https://2ality.com/2020/06/type-guards-assertion-functions-typescript.html#using-overloading for the switch-based narrowing

Reply:
A while back I managed to do a similar recursive type in https://github.com/ocapn/ocapn/blob/dc-codec-play/packages/ocapn-value/src/ocapn-value.d.ts

I wish I had cited a source for the "gymnastics"